processing/src/main/java/io/druid/query/topn/BaseTopNAlgorithm.java

```diff
@@ -74,7 +74,10 @@ public void run(
   {
     boolean hasDimValSelector = (dimValSelector != null);
 
-    final int cardinality = params.getCardinality();
+    int cardinality = params.getCardinality();
```

Comment:
I've changed BaseTopNAlgorithm to have a "with cardinality" path and "no cardinality" path, the "no cardinality" path doesn't use `DimValSelector` (current implementations need cardinality) and the multi-pass logic